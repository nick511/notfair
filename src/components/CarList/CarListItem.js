import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import cn from 'classnames'

import PriceChart from './PriceChart'
import styles from './CarListItem.module.scss'

class CarListItem extends PureComponent {
  _favoriteCar = e => {
    e.preventDefault()
    const { favoriteCarAction, car } = this.props
    favoriteCarAction(car.id, !car.isFavorite)
  }

  render() {
    const { car } = this.props

    const payment = Array.isArray(car.product_financials)
      ? car.product_financials[0]
      : null

    return (
      <Link to={`/vehicles/${car.id}`} className={styles.component}>
        {/* Car image */}
        <div className={styles.image}>
          <img src={car.chrome_image_url} alt="" />
        </div>

        {/* Car info */}
        <div className={styles.info}>
          <div className={styles.name}>
            {car.model_year} {car.make} {car.model}
          </div>

          {payment && (
            <div>
              <span className={styles.price}>
                ${payment.monthly_payment_cents / 100}
              </span>
              per mo.
            </div>
          )}

          <div className={styles.infoItem}>TRIM: {car.trim}</div>
          <div className={styles.infoItem}>MILES: {car.mileage}</div>

          {payment && (
            <div className={styles.infoItem}>
              Start Fee: ${payment.start_fee_cents / 100}
            </div>
          )}

          {/* Favorite button */}
          <div
            className={cn(styles.favorite, 'favorite', {
              favorited: car.isFavorite,
            })}
            onClick={this._favoriteCar}
          >
            <i className="fa fa-heart" />
          </div>
        </div>

        <div className={styles.priceChart}>
          <PriceChart
            realPrice={16999}
            minPrice={12680}
            listPrice={17999}
            fairMaxPrice={17750}
          />
        </div>
      </Link>
    )
  }
}

CarListItem.propTypes = {
  car: PropTypes.shape({
    id: PropTypes.string.isRequired,
    make: PropTypes.string.isRequired,
    model: PropTypes.string.isRequired,
  }).isRequired,
  favoriteCarAction: PropTypes.func.isRequired,
}

export default CarListItem
