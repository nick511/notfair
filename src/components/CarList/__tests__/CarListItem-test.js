import React from 'react'
import { shallow } from 'enzyme'

import CarListItem from '../CarListItem'

describe('<CarListItem />', () => {
  let wrapper = null
  const car = {
    id: '19XFC2F59GE2276732016',
    make: 'Honda',
    mileage: 35292,
    model: 'Civic Sedan',
    model_year: '2016',
    chrome_image_url:
      'http://cdn-prod.prod.fair.engineering/vehicle-images/0640_001_png/MY2016/10846/10846_cc0640_001_BS.png',
    product_financials: [
      {
        id: 19882216,
        monthly_payment_cents: 29000,
        start_fee_cents: 90000,
      },
    ],
    trim: 'LX',
    isFavorite: false,
  }
  const favoriteCarAction = jest.fn()

  beforeEach(() => {
    wrapper = shallow(
      <CarListItem car={car} favoriteCarAction={favoriteCarAction} />,
    )
  })

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders car image', () => {
    expect(wrapper.find('.image img').prop('src')).toEqual(car.chrome_image_url)
  })

  it('renders car name and price', () => {
    expect(wrapper.find('.info .name').text()).toEqual(
      `${car.model_year} ${car.make} ${car.model}`,
    )
    expect(wrapper.find('.info .price').text()).toEqual(
      '$' + car.product_financials[0].monthly_payment_cents / 100,
    )
  })

  it('chages favorite icon style after click the icon', () => {
    const action = () => {
      wrapper.setProps({
        car: { ...car, isFavorite: true },
      })
    }
    wrapper = shallow(<CarListItem car={car} favoriteCarAction={action} />)

    // Should be gray
    expect(wrapper.find('.favorite').hasClass('favorited')).toEqual(false)

    // Should has color
    const event = Object.assign(jest.fn(), { preventDefault: () => {} })
    wrapper.find('.favorite').simulate('click', event)
    expect(wrapper.find('.favorite').hasClass('favorited')).toEqual(true)
  })

  it('dont render payment info if these is no product_financials', () => {
    const testCar = { ...car, product_financials: null }
    wrapper = shallow(
      <CarListItem car={testCar} favoriteCarAction={favoriteCarAction} />,
    )

    expect(wrapper.find('.info .price')).toHaveLength(0)
    expect(wrapper.find('.info .price')).toHaveLength(0)
  })
})
