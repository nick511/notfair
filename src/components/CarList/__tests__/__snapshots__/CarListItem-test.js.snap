// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarListItem /> renders without crashing 1`] = `
<Link
  className="component"
  replace={false}
  to="/vehicles/19XFC2F59GE2276732016"
>
  <div
    className="image"
  >
    <img
      alt=""
      src="http://cdn-prod.prod.fair.engineering/vehicle-images/0640_001_png/MY2016/10846/10846_cc0640_001_BS.png"
    />
  </div>
  <div
    className="info"
  >
    <div
      className="name"
    >
      2016
       
      Honda
       
      Civic Sedan
    </div>
    <div>
      <span
        className="price"
      >
        $
        290
      </span>
      per mo.
    </div>
    <div
      className="infoItem"
    >
      TRIM: 
      LX
    </div>
    <div
      className="infoItem"
    >
      MILES: 
      35292
    </div>
    <div
      className="infoItem"
    >
      Start Fee: $
      900
    </div>
    <div
      className="favorite favorite"
      onClick={[Function]}
    >
      <i
        className="fa fa-heart"
      />
    </div>
  </div>
  <div
    className="priceChart"
  >
    <PriceChart
      fairMaxPrice={17750}
      listPrice={17999}
      minPrice={12680}
      realPrice={16999}
    />
  </div>
</Link>
`;
